Implement memory-efficient model cache

lstein opened the pull request:
Manage a RAM cache of diffusion/transformer models for fast switching. They are moved between GPU VRAM and CPU RAM as necessary. If the cache grows larger than a preset maximum, then the least recently used model will be cleared and (re)loaded from disk when next needed.

## GPU management

The cache acts as a context manager that will load the model into the GPU within the context, and unload it outside the
context. Use like this:

```
cache = ModelCache(max_models=6)
with cache.get_model('runwayml/stable-diffusion-1-5') as SD1,\
     cache.get_model('stabilityai/stable-diffusion-2') as SD2:
          do_something_in_GPU(SD1,SD2)
```
Technically, get_model() returns a context manager Generator, which you can pass around to other functions:
```
sd1_context = cache.get_model('runwayml/stable-diffusion-1-5')
my_function(sd1_context)

def my_function(model_context):
   with model_context as m:
        do_something_with_torch(m)
```

## Manage any model class with `from_pretrained()`

Any type of model that has a `from_pretrained()` method can be loaded and cached with this class, allowing it to manage diffusers pipelines, VAEs, text encoders, etc. You can load a HuggingFace model using its repo_id, or a local model using its path. Legacy checkpoint .ckpt and .safetensors are converted into diffusers on the fly. Precision and the GPU type are set at cache initialization time to ensure that all models are compatible. You can also set `sequential_offload` to minimize GPU VRAM during execution of diffusers pipelines.

```
cache = ModelCache(precision=torch.float32, execution_device=torch.device('mps'), sequential_offload=True)
with cache.get_model('/usr/local/models/sd-1.5') as sd1:
   pass
```
The class of the model returned is controlled by the parameter `model_type`, which defaults to `StableDiffusionGeneratorPipeline`. For convenience, the module provides an Enum named `SDModelType` that enumerates all the model types used in a diffusers pipeline. You can subclass this or use your own Enum to add support for additional model classes.
```
with cache.get_model('openai/clip-vit-large-patch14', model_type=SDModelType.text_encoder) as clip:
   pass
```
## Load parts of models

You can selectively load a portion of a diffusers pipeline, such as its VAE or text encoder. If you know the name of the subfolder in which the part is located, you can load just that bit of the pipeline using the `subfolder` parameter:
```
with cache.get_model('stabilityai/stable-diffusion-2',
                     model_type=SDModelType.tokenizer,
                     subfolder='tokenizer') as tokenizer:
     pass
```
Alternatively, you can retrieve the entire pipeline, and return just the desired part of the model. This is useful if the parent pipeline is already loaded into RAM, or if you need multiple parts from it:
```
with cache.get_model('stabilityai/stable-diffusion-2',
                      submodel=SDModelType.vae) as vae:
   pass
```

## Support for legacy checkpoints

You can pass the path to a legacy checkpoint file (.ckpt or .safetensors) to `get_model()` and it will be converted into a diffusers model in memory. For this to work properly, you need to provide an initialized `LegacyInfo` object which provides the path to the configuration file for this checkpoint and optionally a VAE model to swap in:
```
legacy_data = LegacyInfo(config_file='./configs/v1-inference.yaml')
context = cache.get_model('/path/to/ckpt_file.ckpt',legacy_info=legacy_data)
with context as legacy_model:
    do_something(legacy_model)
```

## Misc

This class can calculate, cache and return unique hashes for diffusers repo_ids, models on disk, and legacy models. In the case of errors it propagates the error to the caller. It also provides a `scan_model()` class method which does pickle scanning on legacy checkpoints, and will raise `UnsafeModelException` and `UnscannableModelException` exceptions in the case of problems.

To write the pytest for this module, I created a dummy pipeline class that implements the minimum methods needed to be managed by the ModelCache. This might be useful for other tests. Please see `tests/test_model_cache.py`

## Relationship to the ModelManager class

I am currently modifying the ModelManager class to utilize the ModelCache and its various features. The revised ModelManager will run as a nodes service, and will be responsible for translating symbolic names in `models.yaml` into their repo_ids, versions and subfolders for passing to ModelCache.

The ModelManager work will appear in a subsequent PR.

psychedelicious commented:
What if the memory management for models was generalized such that, if I attempt to do some memory intensive task (say a large vae decode), if I'm about to run out of memory and there is a model that could be released, the manager releases it?

lstein commented:
> What if the memory management for models was generalized such that, if I attempt to do some memory intensive task (say a large vae decode), if I'm about to run out of memory and there is a model that could be released, the manager releases it?

Sort of. The manager keeps a lock set of the models that are in current use (as defined by the context). When the model goes out of context it isn't immediately removed from the GPU, but stays there in case it is needed soon, which I think will be a common circumstance when a node graph is being executed. If a new model is requested, then **all** unlocked models will be oved out of the GPU in order to free space. So the current limitation is that the manager may purge a model from VRAM when it isn't necessary to do so, not that it holds onto a model longer than it should.

Moving models between RAM and VRAM is pretty fast, but not instantaneous. I would like to be more intelligent about VRAM memory management, but to do this right the manager has to know the size of each model. Earlier, I experimented with various Python modules that walk through objects and sum up their size, but when you apply them to a 4 gig diffusers model they take a long time to run (tens of seconds) and performance became an issue. Maybe the thing to do is to compute the model size when it is loaded from the very first time and store that info to disk.

Unless you feel this is a must-have feature, I'll work on the rest of the model manager service first, and then get back to optimizing VRAM management later.

psychedelicious commented:
Definitely not must have, just an idea :)

StAlKeR7779 commented on invokeai/backend/model_management/model_manager.py:
```diff
@@ -982,9 +1054,9 @@ def preamble(self) -> str:
+        if full_name in self.config:
             return full_name
-        if self.valid_model(model_name):
+        if model_name in self.config:
```
Didn't it falsely triggered when you have for example lora and diffusers with same name?
So, you call `valid_model(lora_name, Type.Lora)` which calls this function and if you have something else with same name as `lora_name` then it will return it like this is lora.
IMHO - we should always use key for config with type and name. It's easier on download to concat type with name then tries to handle different formats of key.

lstein replied:
I want to be able to handle legacy `models.yaml`. I have just committed a fix that allows the bare name to be used if and only if the type requested is 'diffusers' and the stanza's `format` is `diffusers` as well.

The model manager `import` methods all write out stanzas using the `name/type` syntax, so new `models.yaml` will be correct.

lstein commented:
> Definitely not must have, just an idea :)

This is actually now implemented.

psychedelicious commented:
For future consideration #3391

lstein commented:
@StAlKeR7779 I have added the code to convert a legacy VAE .ckpt file into a diffusers model. You'll find it in model_manager under `convert_vae_ckpt_and_cache()`

It seems that the current version no longer supports loading an entire diffusers pipeline. When I try, I get the following stack trace:
```
a = mgr.get_model('stable-diffusion-1.5',SDModelType.Diffusers)
>> Loading model runwayml/stable-diffusion-v1-5, type diffusers:None
╭───────────────────── Traceback (most recent call last) ──────────────────────╮
│ in <module>:1                                                                │
│                                                                              │
│ /home/lstein/Projects/InvokeAI-main/invokeai/backend/model_management/model_ │
│ manager.py:354 in get_model                                                  │
│                                                                              │
│    351 │   │   if model_type == SDModelType.Vae and "vae" in mconfig:        │
│    352 │   │   │   print("NOT_IMPLEMENTED - RETURN CUSTOM VAE")              │
│    353 │   │                                                                 │
│ ❱  354 │   │   model_context = self.cache.get_model(                         │
│    355 │   │   │   location,                                                 │
│    356 │   │   │   model_type = model_type,                                  │
│    357 │   │   │   revision = revision,                                      │
│                                                                              │
│ /home/lstein/Projects/InvokeAI-main/invokeai/backend/model_management/model_ │
│ cache.py:579 in get_model                                                    │
│                                                                              │
│   576 │   │   │                                                              │
│   577 │   │   │   # clean memory to make MemoryUsage() more accurate         │
│   578 │   │   │   gc.collect()                                               │
│ ❱ 579 │   │   │   model_obj = model_info.get_model(submodel, torch_dtype=sel │
│   580 │   │   │   if mem_used := model_info.get_size(submodel):              │
│   581 │   │   │   │   logger.debug(f'CPU RAM used for load: {(mem_used/GIG): │
│   582 │   │   │   │   self.current_cache_size += mem_used   # increment size │
│                                                                              │
│ /home/lstein/Projects/InvokeAI-main/invokeai/backend/model_management/model_ │
│ cache.py:227 in get_model                                                    │
│                                                                              │
│   224 │   ):                                                                 │
│   225 │   │   # return pipeline in different function to pass more arguments │
│   226 │   │   if child_type is None:                                         │
│ ❱ 227 │   │   │   raise Exception("Child model type can't be null on diffuse │
│   228 │   │   if child_type not in self.child_types:                         │
│   229 │   │   │   return None # TODO: or raise                               │
│   230                                                                        │
╰──────────────────────────────────────────────────────────────────────────────╯
Exception: Child model type can't be null on diffusers model
```

Since the code no longer works, I've converted it into a draft.

lstein commented:
@StAlKeR7779 I have merged main into this PR in order to resolve the changes introduced by the revised global configuration management system. pyflakes on `model_cache` is giving me a few undefined names, which I think were there in your earlier commit. Could you check?

```
$ pyflakes model_cache.py
model_cache.py:89:48: undefined name 'index_file'
model_cache.py:90:30: undefined name 'json'
model_cache.py:288:62: undefined name 'repo_id_or_path'
model_cache.py:296:36: undefined name 'transformers'
model_cache.py:308:39: undefined name 'transformers'
model_cache.py:313:75: undefined name 'repo_id_or_path'
model_cache.py:319:70: undefined name 'repo_id_or_path'

Compilation exited abnormally with code 1 at Thu May 18 14:28:04
```

psychedelicious commented:
Getting an error about the safety checker in the compel node:
```python
"Traceback (most recent call last):
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/services/processor.py", line 70, in __process
    outputs = invocation.invoke(
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/invocations/compel.py", line 63, in invoke
    tokenizer_info = context.services.model_manager.get_model(
  File "/home/bat/Documents/Code/InvokeAI/invokeai/app/services/model_manager_service.py", line 338, in get_model
    model_info = self.mgr.get_model(
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/model_management/model_manager.py", line 354, in get_model
    model_context = self.cache.get_model(
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/model_management/model_cache.py", line 630, in get_model
    model_info = self._get_model_info(
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/model_management/model_cache.py", line 612, in _get_model_info
    self.model_infos[model_info_key] = MODEL_TYPES[model_type](
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/model_management/model_cache.py", line 212, in __init__
    self.child_sizes[child_name] = calc_model_size_by_fs(repo_id_or_path, subfolder=child_name)
  File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/model_management/model_cache.py", line 66, in calc_model_size_by_fs
    all_files = os.listdir(model_path)
FileNotFoundError: [Errno 2] No such file or directory: '/home/bat/invokeai/models/hub/models--XpucT--Deliberate/snapshots/5760fbc712b0a637025b62ed60862feeecae2a58/safety_checker'
"
```

The features stanza in `invokeai.yaml` looks like this:
```yaml
  Features:
    esrgan: true
    internet_available: true
    log_tokenization: false
    nsfw_checker: false
    patchmatch: true
    restore: true
```

Graph:
![image](https://github.com/invoke-ai/InvokeAI/assets/4822129/763630b8-40b6-48a5-8ba6-940de7b6db13)

lstein commented:
> Getting an error about the safety checker in the compel node (issue occurs on all three models I tried: `stable-diffusion-1.5`, `deliberate_v1`, and `dreamlike-photoreal-2.0`):
> 
> ```python
> "Traceback (most recent call last):
>   File "/home/bat/Documents/Code/InvokeAI/invokeai/app/services/processor.py", line 70, in __process
>     outputs = invocation.invoke(
>   File "/home/bat/Documents/Code/InvokeAI/invokeai/app/invocations/compel.py", line 63, in invoke
>     tokenizer_info = context.services.model_manager.get_model(
>   File "/home/bat/Documents/Code/InvokeAI/invokeai/app/services/model_manager_service.py", line 338, in get_model
>     model_info = self.mgr.get_model(
>   File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/model_management/model_manager.py", line 354, in get_model
>     model_context = self.cache.get_model(
>   File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/model_management/model_cache.py", line 630, in get_model
>     model_info = self._get_model_info(
>   File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/model_management/model_cache.py", line 612, in _get_model_info
>     self.model_infos[model_info_key] = MODEL_TYPES[model_type](
>   File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/model_management/model_cache.py", line 212, in __init__
>     self.child_sizes[child_name] = calc_model_size_by_fs(repo_id_or_path, subfolder=child_name)
>   File "/home/bat/Documents/Code/InvokeAI/invokeai/backend/model_management/model_cache.py", line 66, in calc_model_size_by_fs
>     all_files = os.listdir(model_path)
> FileNotFoundError: [Errno 2] No such file or directory: '/home/bat/invokeai/models/hub/models--XpucT--Deliberate/snapshots/5760fbc712b0a637025b62ed60862feeecae2a58/safety_checker'
> "
> ```
> 
> The features stanza in `invokeai.yaml` looks like this (changing the `false` to `true` for `nsfw_checker` does not fix the issue):
> 
> ```yaml
>   Features:
>     esrgan: true
>     internet_available: true
>     log_tokenization: false
>     nsfw_checker: false
>     patchmatch: true
>     restore: true
> ```
> 
This should be working now.

ebr commented on invokeai/backend/model_management/model_manager.py:
```diff
@@ -1,418 +1,633 @@
+        #if "__meta__" not in config:
+        #    config = self._convert_2_3_models(config)
+
+        config_meta = ConfigMeta(**config.pop("__metadata__")) # TODO: naming
```
Following @StAlKeR7779 's advice I used the `migrate_models_to_3.0.py ` script to download models and create the new `models.yaml` for 3.0.0, which worked to that end, but now I am hitting this (in both `invokeai-model-install` and `invokeai-configure`: 

```
  File "/home/clipclop/Code/invokeai/invokeai/frontend/install/model_install.py", line 100, in create
    self.installed_cn_models = model_manager.list_controlnet_models()
AttributeError: 'ModelManager' object has no attribute 'list_controlnet_models'
``
